test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -490,79 +565,136 @@ class test_ucp_tag_nbx : public test_ucp_tag {
+        ucp::data_type_desc_t send_dt(datatype, send_buffer, size, iov_count);
+
+        /* Currently recv side supports only contig datatype */
+        ucp::data_type_desc_t recv_dt(ucp_dt_make_contig(1), recv_buffer, size);
+
+        send_param.datatype      = datatype;
+        send_param.op_attr_mask |= UCP_OP_ATTR_FIELD_DATATYPE;
+
+        if (prereg()) {
+            send_param.op_attr_mask |= UCP_OP_ATTR_FIELD_MEMH;
+            recv_param.op_attr_mask |= UCP_OP_ATTR_FIELD_MEMH;
+            send_param.memh = sender().mem_map((void*)send_buffer, size);
+            recv_param.memh = receiver().mem_map(recv_buffer, size);
+        }
+
+        do_send_recv(send_dt, recv_dt, send_param, recv_param);
+
+        if (prereg() && !is_self()) {
+            /* Not relevant for 'self' because both sender and receiver are the same entity.
+               Must be called before request is freed by free_callback (wait_for_value). */
+            test_prereg_rcache_stats(send_dt, recv_dt, send_param, recv_param);
+        }
 
-        if (prereg) {
+        if (!completion.empty()) {
+            /* Wait for completion indication */
+            wait_for_value(completion.m_address, completion.m_value);
+        }
+
+        if (prereg()) {
             sender().mem_unmap(send_param.memh);
             receiver().mem_unmap(recv_param.memh);
         }
     }
 
-    void test_recv_send(size_t size = MSG_SIZE, bool prereg = false)
+    void test_recv_send(size_t size = MSG_SIZE)
     {
         std::vector<char> send_buffer(size);
         std::vector<char> recv_buffer(size);
-        test_recv_send(size, &send_buffer[0], &recv_buffer[0], prereg);
+        test_recv_send(size, &send_buffer[0], &recv_buffer[0]);
     }
 };
 
 const size_t test_ucp_tag_nbx::MSG_SIZE  = 4 * UCS_KBYTE * ucs_get_page_size();
 const ucp_request_param_t test_ucp_tag_nbx::null_param = {0};
-
+const test_ucp_tag_nbx::completion_value test_ucp_tag_nbx::null_completion =
+        {nullptr, 0};
+const size_t test_ucp_tag_nbx::iov_count               = 20;
 
 UCS_TEST_P(test_ucp_tag_nbx, basic)
 {
     test_recv_send();
 }
 
-UCS_TEST_P(test_ucp_tag_nbx, eager_zcopy_prereg, "ZCOPY_THRESH=0",
-           "RNDV_THRESH=inf")
+UCS_TEST_P(test_ucp_tag_nbx, eager_zcopy, "ZCOPY_THRESH=0", "RNDV_THRESH=inf")
 {
-    test_recv_send(4 * UCS_KBYTE, true);
+    test_recv_send(4 * UCS_KBYTE);
 }
 
-UCS_TEST_P(test_ucp_tag_nbx, rndv_prereg, "RNDV_THRESH=0")
+UCS_TEST_P(test_ucp_tag_nbx, rndv, "RNDV_THRESH=0")
 {
-    test_recv_send(64 * UCS_KBYTE, true);
+    test_recv_send(64 * UCS_KBYTE);
 }
 
-UCS_TEST_P(test_ucp_tag_nbx, fallback)
+UCS_TEST_P(test_ucp_tag_nbx, fallback, "ZCOPY_THRESH=inf")
 {
-    if (m_ucp_config->ctx.proto_enable) {
-        UCS_TEST_SKIP_R("FIXME: fallback is not implemented in proto_v2");
+    if (proto_enbaled() || prereg()) {
+        UCS_TEST_SKIP_R(
+                "protoV2/prereg are not supported for partial md reg failure");
```

Comment:
is it still relevant?

Reply:
they both crash on failed asserts

Reply:
@shasson5 - can elaborate, why prereg fails?

Reply:
`ucp_mem_map` returns bad status, which cause the test to be aborted